rfd/0037-desktop-access-protocol.md

```diff
@@ -0,0 +1,191 @@
+     |--------------------------------------->|
+     |             4 - mouse button           |
+     |--------------------------------------->|
+     |             2 - PNG frame              |
+     |<---------------------------------------|
+     |             5 - keyboard input         |
+     |--------------------------------------->|
+     |             2 - PNG frame              |
+     |<---------------------------------------|
+     |                ....                    |
+```
+
+### Message encoding
+
+Each message consists of a sequence of fields.
+Each field is either fixed size or variable size.
+The first byte in each message is the message type and defines what fields are
+expected after it.
+
+#### Field types
+
+Fields are all numbers, using Go-inspired names. Numbers are encoded in big
+endian order.
+For example:
+- `byte` is a single byte
+- `uint32` is an unsigned 32-bit integer
+- `int64` is a signed 64-bit integer
+
+Message definitions use the syntax `[]type` to declare a variable size field
+with elements of type `type`. The length should be deducted from nearby fields.
+
+Strings are encoded as UTF-8 in a `[]byte` field, with a `uint32` length
+prefix.
+
+### Message types
+
+#### 1 - client screen spec
+
+```
+| message type (1) | width uint32 | height uint32 |
+```
+
+This message contains the dimensions of the client view - the dimensions used
+for drawing the remote desktop image. Sent from client to server.
+
+This message can be sent more than once per session, for example when client
+resizes their window.
+
+#### 2 - PNG frame
+
+```
+| message type (2) | left uint32 | top uint32 | right uint32 | bottom uint32 | data []byte |
+```
+
+This message contains new bitmap data for a region of the desktop screen. Sent
+from server to client.
+
+`left`, `top` and `right`, `bottom` contain the top-left and bottom-right
+coordinates of the region, in pixels.
+`data` contains the PNG-encoded bitmap.
+
+#### 3 - mouse move
+
+```
+| message type (3) | x uint32 | y uint32 |
+```
+
+This message contains new mouse coordinates. Sent from client to server.
+
+#### 4 - mouse button
+
+```
+| message type (4) | button byte | state byte |
+```
+
+This message contains a mouse button update. Sent from client to server.
+
+`button` identifies which button was used:
+- `0` is left button
+- `1` is middle button
+- `2` is right button
+
+`state` identifies the button state:
+- `0` is not pressed
+- `1` is pressed
+
+#### 5 - keyboard input
+
+```
+| message type (5) | key_code uint32 | state byte |
+```
+
+This message contains a keyboard update. Sent from client to server.
+
+`key_code` is the keyboard code from https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code/code_values#code_values_on_windows
+
+`state` identifies the key state:
+- `0` is not pressed
+- `1` is pressed
+
+Key combinations show up as a sequence of keys going into `state=1` and then
+back to `state=0`.
+
+#### 6 - clipboard data
```

Comment:
What's the plan for implementing clipboard on the client (since it's a web app)? Asynchronous Clipboard API?

Reply:
@ibeckermayer was looking into this.

Reply:
Yes, [Asynchronous Clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API)

Reply:
Thanks! 

@phosphore This would be part of the implementation review too.